src/Mod/Sketcher/Gui/TaskSketcherTool.h

```diff
@@ -0,0 +1,113 @@
+/***************************************************************************
+ *   Copyright (c) 2022 Pierre-Louis Boyer <pierrelouis.boyer@gmail.com>   *
+ *                                                                         *
+ *   This file is part of the FreeCAD CAx development system.              *
+ *                                                                         *
+ *   This library is free software; you can redistribute it and/or         *
+ *   modify it under the terms of the GNU Library General Public           *
+ *   License as published by the Free Software Foundation; either          *
+ *   version 2 of the License, or (at your option) any later version.      *
+ *                                                                         *
+ *   This library  is distributed in the hope that it will be useful,      *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU Library General Public License for more details.                  *
+ *                                                                         *
+ *   You should have received a copy of the GNU Library General Public     *
+ *   License along with this library; see the file COPYING.LIB. If not,    *
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#ifndef GUI_TASKVIEW_TaskSketcherTool_H
+#define GUI_TASKVIEW_TaskSketcherTool_H
+
+#include <Gui/TaskView/TaskView.h>
+#include <Gui/TaskView/TaskDialog.h>
+#include <Gui/Selection.h>
+#include <boost_signals2.hpp>
+#include <Inventor/events/SoKeyboardEvent.h>
+
+namespace App {
+class Property;
+}
+
+namespace Gui {
+class ViewProvider;
+}
+
+namespace SketcherGui {
+
+class Ui_TaskSketcherTool;
+class ViewProviderSketch;
+
+class SketcherToolWidget : public QWidget
+{
+    Q_OBJECT
+
+public:
+    SketcherToolWidget(QWidget *parent=0, ViewProviderSketch* sketchView=0);
+    ~SketcherToolWidget();
+
+    bool eventFilter(QObject* object, QEvent* event);
+
+    void loadSettings();
+
+    void setparameter(double val, int i);
+    void setLabel(const QString& atext, int i);
+    void setUnit(Base::Unit unit, int i);
+    void setParameterActive(bool val, int i);
+    void setParameterVisible(bool val, int i);
+    void setParameterFocus(int i);
+    void setSettings(int toolSelected);
+    bool isCheckBoxChecked(int i);
+    bool isWidgetActive;
+    std::vector<double> toolParameters;
+    std::vector<bool> isSettingSet;
+    bool unitChangeTriggeredValueChanged;
+
+//Q_SIGNALS:
+public Q_SLOTS:
+    void emitSetparameter(double val);

```

Comment:
Oh, yes the slots. I do not know exactly how you will want to use these. I have left them in the code, but you may want consider to connect them to other class, or to signal to another class within the slot. 

Code in the default widget should be generic enough to allow any use of parameters.

Reply:
Slots are doing : `isSettingSet[i] = 1;` which let drawSketchHandler if user input a value or not.
How should this be handled now? Should I put back isSettingSet bool vector in the default widget?

Reply:
> Oh, yes the slots. I do not know exactly how you will want to use these. I have left them in the code, but you may want consider to connect them to other class, or to signal to another class within the slot.

They have 4 functions : 1- register that the parameter has been set at least once in the bool vector : isSettingSet.
2- set focus to next spinbox
3- Trigger a mouseMove of the drawsketchHandler. This enable to update the geometry and go to next step (STATUS_SEEK_Second...) automatically if user typed values.
4- When spinbox unit is changed (for angles for instance) valueChanged is triggered, which triggers isSettingSet. So then the DrawSketchHandler thinks we entered a value. Though this might not be an issue when the widget factory is used.

Please advise how 3- should be handled now. Thanks

```
void SketcherToolWidget::emitSetparameter(double val)
{
    isSettingSet[0] = 1;
    toolParameters[0] = val;
    if (unitChangeTriggeredValueChanged) {
        unitChangeTriggeredValueChanged = 0;
    }
    else setParameterFocus(1);

    //Make a mousemove to update geometry when parameter is set.
    if(isWidgetActive) {
        sketchView->mouseMove(sketchView->prvMoveCursorPos, sketchView->prvMoveViewer);
    }
}
```

Reply:
Is it intended that the widget is private to toolWidgetManager?
Then nothing can be used directly from the handler. 
For instance if I want to do 
toolWidget->setParameterFocus(x1);
I need to create a function in toolWidgetManager? What is the point it feels like it just makes things convoluted. Or is it to prepare for widget factory?